### PR TITLE
Feature/swift rewrite

### PR DESCRIPTION
**Pull request by simonmitchell**
This is a complete re-write of the whole project using Swift. It brings slight breaking changes to the API, for example:

GET,POST,DELETE methods have all been moved into a single method `request` which take the HTTP method as an enum parameter

Request body must now be an implementer of the `RequestBody` protocol, so no just sending in `Any` for a request any more!

Plenty of tests have now been added for both responses, request bodies and other parts of the Framework. More should be written, but we have significantly more now than we ever had with the Objective-C version!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,24 +2,190 @@
...
+
+```
+let bodyJSON = [
+    "name": "Thunder Request",
+    "isAwesome": true
+]
+let body = JSONRequestBody(bodyJSON)
+```
+
+#### JSONRequestBody
```
Comment (ryanbourneuk) [inline README.md]: Copy pasta?

File: ThunderRequest/Authenticator.swift [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/// in the completion block to the keychain using `RequestCredential.store(credential:withIdentifier)`.
+/// Once you have stored the credential in the keychain all requests will check that it hasn't expired
+/// before making the request.
+///
+/// Setting the `RequestController`'s `sharedRequestCredential` using `set(sharedRequestCredentials:savingToKeychain:)`
+/// with savingToKeychain as true will also achieve the same affect.
+public protocol Authenticator {
+    
+    /// This method will be called if a request is made without a `RequestCredential` object having
+    /// been saved to the keychain under `authIdentifier`
```
Comment (ryanbourneuk) [inline ThunderRequest/Authenticator.swift]: Is it instead saved into the keychain under whatever `authIdentifier` is defined as? Unsure if it's just me but it feels like it reads that it'll be saved with a key of `authIdentifier` :)

File: ThunderRequest/Authenticator.swift [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    func authenticate(completion: (_ credential: RequestCredential?, _ error: Error?, _ saveToKeychain: Bool) -> Void)
+    
+    /// This defines the service identifier for the auth flow, which the credentials object will
+    /// be saved under in the user's keychain
+    var authIdentifier: String { get }
+    
+    /// The accessibility level of the credential when stored in the user's keychain
+    var keychainAccessibility: CredentialStore.Accessibility { get }
+    
+    /// This method will be called if a request is made with an expired token, or if we recieve a 403 challenge from a particular request
```
Comment (ryanbourneuk) [inline ThunderRequest/Authenticator.swift]: `receive`

File: ThunderRequest/CredentialStore.swift [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    /// - always: The data in the keychain item can always be accessed regardless of whether the device is locked. Migrates to new devices.
+    /// - whenUnlocked: The data in the keychain item can be accessed only while the device is unlocked by the user. Migrates to new devices.
+    /// - whenPasscodeSetThisDeviceOnly: The data in the keychain can only be accessed when the device is unlocked. Only available if a passcode is set on the device. Does not migrate to new devices.
+    /// - whenUnlockedThisDeviceOnly: The data in the keychain item can be accessed only while the device is unlocked by the user. Does not migrate to new devices.
+    /// - afterFirstUnlockThisDeviceOnly: The data in the keychain item cannot be accessed after a restart until the device has been unlocked once by the user. Does not migrate to new devices.
+    /// - alwaysThisDeviceOnly: The data in the keychain item can always be accessed regardless of whether the device is locked. Does not migrate to new devices.
+    public enum Accessibility {
```
Comment (ryanbourneuk) [inline ThunderRequest/CredentialStore.swift:128]: This is really nice!

File: ThunderRequest/CredentialStore.swift [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    ///
+    /// - Parameters:
+    ///   - credential: The credentials object to store in the keychain
+    ///   - identifier: The identifier to store the credential object under
+    ///   - accessibility: The access rule for the credential
+    /// - Returns: Whether the item was sucessfully stored
+    @discardableResult public static func store(credential: RequestCredential?, identifier: String, accessibility: Accessibility = .afterFirstUnlock, in store: DataStore = KeychainStore(serviceName: kTSCAuthServiceName)) -> Bool {
+        
+        guard let credential = credential else {
+            return delete(withIdentifier: identifier)
```
Comment (ryanbourneuk) [inline ThunderRequest/CredentialStore.swift:169]: Would be good to document that passing a nil credential deletes whatever is under the identifier. Perhaps it maybe shouldn't even do this, and instead you should explicitly call the `delete(withIdentifier:...)` function instead of relying on passing `nil` here. What do you think?

File: ThunderRequest/CredentialStore.swift [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        let exists = existingCredential != nil
+        
+        if exists {
+            return store.update(data: credential.keychainData, identifier: identifier, accessibility: accessibility)
+        } else {
+            return store.add(data: credential.keychainData, identifier: identifier, accessibility: accessibility)
+        }
+    }
+    
+    /// Deletes an entry for a certain identifier from the keychain
```
Comment (ryanbourneuk) [inline ThunderRequest/CredentialStore.swift]: Copy pasta?

File: ThunderRequest/CredentialStore.swift [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    ///
+    /// - Parameter withIdentifier: The identifier to delete the credential object for
+    /// - Returns: The retrieved credential
+    /// - Throws: An error if retrieval fails
```
Comment (ryanbourneuk) [inline ThunderRequest/CredentialStore.swift]: Is it throwing an error at the moment? Would be good actually if it could, instead of it returning nil. That way it'd be more explicit that something has gone wrong, what do you think? :)

File: ThunderRequest/CustomisableRecoverableError.swift [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+//  ErrorRecoveryAttempter.swift
+//  ThunderRequest
+//
+//  Created by Simon Mitchell on 14/12/2018.
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import Foundation
+
+/// A protocol that inherits from `RecoverableError` and
```
Comment (ryanbourneuk) [inline ThunderRequest/CustomisableRecoverableError.swift]: Is there meant to be something after the `and`? :)

File: ThunderRequest/Data+ContentType.swift [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+//  ContentType+Data.swift
+//  ThunderRequest
+//
+//  Created by Simon Mitchell on 11/12/2018.
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import Foundation
+
+
```
Comment (ryanbourneuk) [inline ThunderRequest/Data+ContentType.swift]: Lots of whitespace here that could be purged :)

File: ThunderRequest/Data+ContentType.swift [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+            return "txt"
+        case "video/quicktime":
+            return "mov"
+        default:
+            return nil
+        }
+    }
+}
+
+//extension Request.ContentType {
```
Comment (ryanbourneuk) [inline ThunderRequest/Data+ContentType.swift]: Is this meant to be commented out? If so it could perhaps be removed?

File: ThunderRequest/EncodableRequestBody.swift [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+            switch self {
+            case .json:
+                return "application/json"
+            case .plist:
+                return "text/x-xml-plist"
+            }
+        }
+    }
+    
+    let encoding: Encoding
```
Comment (ryanbourneuk) [inline ThunderRequest/EncodableRequestBody.swift:34]: This could do with some documentation :)

File: ThunderRequest/ErrorRecoveryOption.swift [hunk 12]
```diff
@@ -0,0 +1,52 @@
+//
+//  ErrorRecoveryOption.swift
+//  ThunderRequest
+//
+//  Created by Simon Mitchell on 14/12/2018.
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import Foundation
+
```
Comment (ryanbourneuk) [inline ThunderRequest/ErrorRecoveryOption.swift:10]: Whitespace

File: ThunderRequest/ErrorRecoveryOption.swift [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    /// - custom: A custom option for recovering from the error
+    /// - retry: Displays a retry button and repeats the request where possible
+    /// - cancel: Cancels the recovery
+    enum Style {
+        case custom
+        case retry
+        case cancel
+    }
+    
+    public typealias Handler = (_ option: ErrorRecoveryOption, _ callback: ((Bool) -> Void)?) -> Void
```
Comment (ryanbourneuk) [inline ThunderRequest/ErrorRecoveryOption.swift:29]: Docs please :)

File: ThunderRequest/FormURLEncodedRequestBody.swift [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+//  ThunderRequest
+//
+//  Created by Simon Mitchell on 11/12/2018.
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import Foundation
+
+/// A request body struct which can be used to represent the payload of a
+/// form url encoded request
```
Comment (ryanbourneuk) [inline ThunderRequest/FormURLEncodedRequestBody.swift]: Can probably go on the line above, won't make it too much longer

File: ThunderRequest/HTTP.swift [hunk 15]
```diff
@@ -0,0 +1,299 @@
...
+public extension RequestBody {
+    
+    func mutate(url: inout URL) {
+        
+    }
+}
+
+
+
+///// An enum providing standard values for the `Content-Type` header.
```
Comment (ryanbourneuk) [inline ThunderRequest/HTTP.swift]: Feel I've seen this elsewhere in the codebase. Was this moved to elsewhere? If so it can probably be deleted now

File: ThunderRequest/HTTP.swift [hunk 16]
```diff
@@ -0,0 +1,299 @@
...
+    /// - unsupportedMediaType: The request entity has a media type which the server or resource does not support. For example, the client uploads an image as image/svg+xml, but the server requires that images use a different format.
+    /// - rangeNotSatisfiable: The client has asked for a portion of the file (byte serving), but the server cannot supply that portion. For example, if the client asked for a part of the file that lies beyond the end of the file. Called "Requested Range Not Satisfiable" previously.
+    /// - expectationFailed: The server cannot meet the requirements of the Expect request-header field.
+    /// - imATeapot: This code was defined in 1998 as one of the traditional IETF April Fools' jokes, in RFC 2324, Hyper Text Coffee Pot Control Protocol, and is not expected to be implemented by actual HTTP servers. The RFC specifies this code should be returned by teapots requested to brew coffee. This HTTP status is used as an Easter egg in some websites, including Google.com.
```
Comment (ryanbourneuk) [inline ThunderRequest/HTTP.swift:106]: 🍵

File: ThunderRequest/RequestController+Callbacks.swift [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    public static let ErrorDomain = "com.threesidedcube.ThunderRequest"
+    
+    static let DidReceiveResponseNotificationName = Notification.Name(rawValue: "TSCRequestDidReceiveResponse")
+    
+    static let DidErrorNotificationName = Notification.Name("TSCRequestServerError")
+    
+    func add(completionHandler: TransferCompletion?, progressHandler: ProgressHandler?, forTaskId taskId: Int) {
+        
+        if transferCompletionHandlers[taskId] != nil {
+//            os_log_error(request_controller_log, "Error: Got multiple handlers for a single task identifier.  This should not happen.\n");
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController+Callbacks.swift]: Are these meant to be commented out?

File: ThunderRequest/RequestController+SessionDelegate.swift [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+        callTransferCompletionHandlersFor(taskIdentifier: downloadTask.taskIdentifier, downloadedFileURL: location, error: nil, response: downloadTask.response)
+    }
+    
+    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didWriteData bytesWritten: Int64, totalBytesWritten: Int64, totalBytesExpectedToWrite: Int64) {
+        let progress = Double(totalBytesWritten)/Double(totalBytesExpectedToWrite)
+        callProgressHandlerFor(taskIdentifier: downloadTask.taskIdentifier, progress: progress, totalBytes: totalBytesExpectedToWrite, progressBytes: totalBytesWritten)
+    }
+    
+    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didResumeAtOffset fileOffset: Int64, expectedTotalBytes: Int64) {
+
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController+SessionDelegate.swift:23]: Is this meant to be empty?

File: ThunderRequest/RequestController+SessionDelegate.swift [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+        let progress = Double(totalBytesSent)/Double(totalBytesExpectedToSend)
+        callProgressHandlerFor(taskIdentifier: task.taskIdentifier, progress: progress, totalBytes: totalBytesExpectedToSend, progressBytes: totalBytesSent)
+    }
+    
+    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
+        callTransferCompletionHandlersFor(taskIdentifier: task.taskIdentifier, downloadedFileURL: nil, error: error, response: task.response)
+    }
+    
+    func urlSession(_ session: URLSession, task: URLSessionTask, willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest, completionHandler: @escaping (URLRequest?) -> Void) {
+
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController+SessionDelegate.swift:36]: Whitespace

File: ThunderRequest/RequestController.swift [hunk 20]
```diff
@@ -0,0 +1,727 @@
...
+//
+
+import Foundation
+import os.log
+
+public typealias RequestCompletion = (_ response: RequestResponse?, _ error: Error?) -> Void
+public typealias TransferCompletion = (_ response: RequestResponse?, _ fileLocation: URL?, _ error: Error?) -> Void
+public typealias ProgressHandler = (_ progress: Double, _ totalBytes: Int64, _ transferredBytes: Int64) -> Void
+
+/// An instance of `RequestController` lets you asynchronously perform HTTP requests with a closure being called upn completion.
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController.swift]: `upon` instead of `upn`

File: ThunderRequest/RequestController.swift [hunk 21]
```diff
@@ -0,0 +1,727 @@
...
+//  ThunderRequest
+//
+//  Created by Simon Mitchell on 11/12/2018.
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import Foundation
+import os.log
+
+public typealias RequestCompletion = (_ response: RequestResponse?, _ error: Error?) -> Void
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController.swift:12]: Loving the use of typealias' :D

File: ThunderRequest/RequestController.swift [hunk 22]
```diff
@@ -0,0 +1,727 @@
...
+    
+    ///MARK: - Initialization -
+    
+    internal let dataStore: DataStore
+    
+    internal init(baseURL: URL, dataStore: DataStore) {
+        
+        self.dataStore = dataStore
+        
+        if baseURL.absoluteString.hasSuffix("/") {
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController.swift:111]: What is the purpose of this? Not sure if I've seen discussions in some places where we may wish to not append `/` to certain URLs for certain app APIs. Perhaps it should be the responsibility of the code calling this to ensure the URL is correctly formed, and not have this here?

File: ThunderRequest/RequestController.swift [hunk 23]
```diff
@@ -0,0 +1,727 @@
...
+        if let userAgent = userAgent {
+            UserDefaults.standard.set(userAgent, forKey: "TSCUserAgent")
+        } else {
+            UserDefaults.standard.removeObject(forKey: "TSCUserAgent")
+        }
+    }
+    
+    private var sessionDelegate: SessionDelegateProxy?
+    
+    var transferCompletionHandlers: [Int : TransferCompletion] = [:]
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController.swift:150]: Some occurrences of spaces before colons in property declarations in this PR. Won't comment everywhere (+ it isn't a huge thing!) but would be good to make it consistent!

File: ThunderRequest/RequestController.swift [hunk 24]
```diff
@@ -0,0 +1,727 @@
...
+    }
+    
+    //MARK: - Making Requests -
+    //MARK: General Requests
+    
+    private func setHeaders(_ headers: [String: String?]?, for request: Request) {
+        
+        var allHeaders = sharedRequestHeaders
+        
+        // In some APIs an error will be returned if you set a Content-Type header
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController.swift:189]: Could be good if this formed part of the docs for this function? Then again, the function is private and this is already super clear, so your call!

File: ThunderRequest/RequestController.swift [hunk 25]
```diff
@@ -0,0 +1,727 @@
...
+    @discardableResult public func request(
+        _ path: String?,
+        method: HTTP.Method,
+        body: RequestBody? = nil,
+        tag: Int = Int.random(in: 0...1000),
+        contentType: String? = nil,
+        overrideURL: URL? = nil,
+        queryItems: [URLQueryItem]? = nil,
+        headers: [String: String?]? = nil,
+        completion: RequestCompletion?) -> Request {
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController.swift]: Does completion also need a default value of nil as well, seeing as it is specified in the docs as optional?

File: ThunderRequest/RequestController.swift [hunk 26]
```diff
@@ -0,0 +1,727 @@
...
+    /// - Returns: The request which has been made
+    @discardableResult public func uploadFile(
+        _ fileURL: URL,
+        to path: String?,
+        tag: Int = Int.random(in: 0...1000),
+        contentType: String? = nil,
+        overrideURL: URL? = nil,
+        queryItems: [URLQueryItem]? = nil,
+        headers: [String: String?]? = nil,
+        progress: ProgressHandler?,
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController.swift]: Same for these as well, should they have default values specified?

Comment (ryanbourneuk) [reply ThunderRequest/RequestController.swift]: Same for the other functions in this file, too :)

File: ThunderRequest/RequestCredential.swift [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+    }
+    
+    /// Initialises a new OAuth2 credential with given parameters
+    ///
+    /// - Parameters:
+    ///   - authorizationToken: The authorizationToken to be sent by `RequestController` for authentication requests.
+    ///   - refreshToken: The refresh token to be sent back to the authenticating endpoint for certain authentification methods.
+    ///   - expiryDate: The date upon which the credential will expire for the user.
+    ///   - tokenType: The token type of the credential (Defaults to Bearer)
+    public init(authorizationToken: String, refreshToken: String?, expiryDate: Date, tokenType: String = "Bearer") {
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestCredential.swift:96]: If the default is `Bearer`, it could be good to move it to a static and then use that static property here and on line 30 as well?

File: ThunderRequest/RequestResponse.swift [hunk 28]
```diff
@@ -0,0 +1,83 @@
...
+    
+    /// Raw data returned from the server
+    public let data: Data?
+    
+    /// The `HTTPURLResponse` object returned from the request. Contains info such as the response code.
+    public let httpResponse: HTTPURLResponse?
+    
+    /// The redirect response
+    /// If the request was redirected, this represents the URLResponse for the original request
+    public var redirectResponse: HTTPURLResponse?
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestResponse.swift]: Maybe this should be instead called `redirectedResponse` if it is the URLResponse that has been redirected from? Lemme know your thoughts!

File: ThunderRequest/UIAlertController+ErrorRecovery.swift [hunk 29]
```diff
@@ -0,0 +1,109 @@
...
+    
+    /// Returns a summarised message body to display to the user combining
+    /// failure reasons and suggested recovery options if supplied
+    ///
+    /// - Parameter error: The error to format for
+    /// - Returns: An optional message
+    static func recoveryMessageFor(error: Error) -> String? {
+        
+        let recoverableError = error as? CustomisableRecoverableError
+        let override = ErrorOverrides.overrideFor(domain: recoverableError?.domain ?? (error as NSError).domain, code: recoverableError?.code ?? (error as NSError).code)
```
Comment (ryanbourneuk) [inline ThunderRequest/UIAlertController+ErrorRecovery.swift:32]: Oooooh, didn't realise the term `override` could be used! Thought it was reserved but seems it's actually fine! :D

File: ThunderRequest/URLRequest+TaskIdentifier.swift [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import Foundation
+
+private var identifierKey: UInt8 = 0
+
+extension URLRequest {
+    
+    var taskIdentifier: Int? {
```
Comment (ryanbourneuk) [inline ThunderRequest/URLRequest+TaskIdentifier.swift:17]: Would be good to get docs here!

File: ThunderRequest/URLSessionTask+Tag.swift [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import Foundation
+
+private var tagKey: UInt8 = 0
+
+extension URLSessionTask {
+    
+    var tag: Int? {
```
Comment (ryanbourneuk) [inline ThunderRequest/URLSessionTask+Tag.swift:17]: Docs here too!

File: ThunderRequest/UIAlertController+ErrorRecovery.swift (hunk 29 above)
Comment (simonmitchell) [reply ThunderRequest/UIAlertController+ErrorRecovery.swift:32]: Seems to be! Looks like it's just on GitHub that it's been picked up by syntax highlighting :P

File: ThunderRequest/RequestResponse.swift (hunk 28 above)
Comment (simonmitchell) [reply ThunderRequest/RequestResponse.swift]: I'm unsure on this one... `redirected` says to me that it's post re-direction (which it isn't) but I also agree that `redirect` isn't perhaps the best term either... perhaps it's just made clearer by the docs?

File: .travis.yml [hunk 32]
```diff
@@ -1,33 +1,33 @@
-language: objective-c
+language: swift
 xcode_project: ThunderRequest.xcodeproj # path to your xcodeproj folder
 xcode_scheme: ThunderRequest
-osx_image: xcode9.3
+osx_image: xcode10.1
```
Comment (ryanbourneuk) [inline .travis.yml:4]: Won't make any more notes in this PR as to not overfill it, but after this it'd be good to get Travis running / check everything is fine with Xcode 10.2. Also any Xcode 10.2 work (think it's already done though!) can go in another PR too. Will be the same for other framework PRs!

File: ThunderRequest/CredentialStore.swift [hunk 33]
```diff
@@ -0,0 +1,174 @@
...
+        var query = keychainQueryWith(identifier: identifier, accessibility: accessibility)
+        query[kSecValueData] = data
+        let status = SecItemAdd(query as CFDictionary, nil)
+        return status == errSecSuccess
+    }
+    
+    public func update(data: Data, identifier: String, accessibility: CredentialStore.Accessibility) -> Bool {
+        
+        // Send nil here because if we send accessibility we get an -25300 status code (errSecItemNotFound)
+        let query = keychainQueryWith(identifier: identifier, accessibility: nil)
```
Comment (ryanbourneuk) [inline ThunderRequest/CredentialStore.swift:82]: So would that mean that the accessibility parameter isn't used? Could there be any side effects to this? :)

File: ThunderRequest/CredentialStore.swift [hunk 34]
```diff
@@ -0,0 +1,174 @@
...
+        
+        return result as? Data
+    }
+    
+    public func removeDataFor(identifier: String) -> Bool {
+        let result = SecItemDelete(keychainQueryWith(identifier: identifier) as CFDictionary)
+        return result == errSecSuccess
+    }
+    
+    public let serviceIdentifier: String
```
Comment (ryanbourneuk) [inline ThunderRequest/CredentialStore.swift]: Would be nice if this and the init function was at the top of the function. Not too fussed on it though :)

File: ThunderRequest/Data+ContentType.swift (hunk 9 above)
Comment (ryanbourneuk) [reply ThunderRequest/Data+ContentType.swift]: Actually, we could always resolve things like this with automated tools. Will leave that for another PR though!

File: ThunderRequest/Data+MultipartFormElement.swift [hunk 35]
```diff
@@ -0,0 +1,30 @@
...
+//
+//  Created by Simon Mitchell on 11/12/2018.
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import Foundation
+
+extension Data: MultipartFormElement {
+    
+    public func multipartDataWith(boundary: String, key: String) -> Data? {
```
Comment (ryanbourneuk) [inline ThunderRequest/Data+MultipartFormElement.swift:13]: Docs please! :)

File: ThunderRequest/HTTP.swift [hunk 36]
```diff
@@ -0,0 +1,299 @@
...
+//}
+
+public struct HTTP {
+    /// Enum representing HTTP Methods
+    ///
+    /// - CONNECT: The CONNECT method establishes a tunnel to the server identified by the target resource.
+    /// - DELETE: The DELETE method deletes the specified resource.
+    /// - GET: The GET method requests a representation of the specified resource. Requests using GET should only retrieve data.
+    /// - HEAD: The HEAD method asks for a response identical to that of a GET request, but without the response body.
+    /// - OPTIOSN: The OPTIONS method is used to describe the communication options for the target resource.
```
Comment (ryanbourneuk) [inline ThunderRequest/HTTP.swift]: `OPTIONS` instead of `OPTIOSN`

File: ThunderRequest/RequestController+Auth.swift [hunk 37]
```diff
@@ -0,0 +1,75 @@
...
+    
+    /// Checks the authentication status for a given request
+    ///
+    /// - Parameters:
+    ///   - request: The request to check authentication for
+    ///   - completion: Closure callback with result
+    func checkAuthStatusFor(request: Request, completion: @escaping AuthCheckCompletion) {
+        
+        guard let authenticator = authenticator else {
+            completion(true, nil, false)
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController+Auth.swift:23]: So if there's no authenticator, we're deemed to be 'authenticated'? Thinking that perhaps this should return an error instead, as surely if we're looking to auth we need to have an authenticator and not having one would be an issue? Unless there's some legacy reason for this?

File: ThunderRequest/RequestController+Callbacks.swift [hunk 38]
```diff
@@ -0,0 +1,206 @@
...
+
+import Foundation
+import os.log
+
+public struct RequestNotificationKey {
+    public static let request = "TSCRequestNotificationRequestKey"
+    public static let response = "TSCRequestNotificationResponseKey"
+}
+
+public extension HTTP {
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController+Callbacks.swift]: Should this be in this file? Thinking HTTP+Error.swift may be better?

File: ThunderRequest/RequestController+Callbacks.swift [hunk 39]
```diff
@@ -0,0 +1,206 @@
...
+        
+        NotificationCenter.default.post(name: RequestController.DidReceiveResponseNotificationName, object: nil, userInfo: requestInfo)
+        
+        if response?.status.isConsideredError == true {
+            NotificationCenter.default.post(name: RequestController.DidErrorNotificationName, object: nil, userInfo: requestInfo)
+        }
+        
+        defer {
+            
+            if let error = error {
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController+Callbacks.swift]: Can this sit in its own function? :)

File: ThunderRequest/RequestController.swift [hunk 40]
```diff
@@ -0,0 +1,736 @@
...
+        guard savingToKeychain else { return }
+        CredentialStore.store(credential: sharedRequestCredentials, identifier: authenticator?.authIdentifier ?? "thundertable.com.threesidedcube-\(sharedBaseURL)", accessibility: accessibility)
+    }
+}
+
+extension RequestController {
+    
+    fileprivate static func showApplicationActivityIndicator() {
+        #if os(iOS)
+        if let option = Bundle.main.object(forInfoDictionaryKey: "TSCThunderRequestShouldHideActivityIndicator") as? Bool, !option {
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestController.swift]: Just double checking this, is this definitely the right key? Also, would be good to move this to a constant :)

File: ThunderRequest/RequestCredential.swift [hunk 41]
```diff
@@ -0,0 +1,133 @@
...
+    /// Creates a new auth token based credential
+    ///
+    /// - Parameter authorizationToken: The authorization token to use
+    init(authorizationToken: String) {
+        super.init()
+        self.authorizationToken = authorizationToken
+    }
+    
+    public func encode(with aCoder: NSCoder) {
+        aCoder.encode(username, forKey: "username")
```
Comment (ryanbourneuk) [inline ThunderRequest/RequestCredential.swift]: Hardcoded keys here that'd be nice to move :)

File: ThunderRequest/RequestResponse.swift (hunk 28 above)
Comment (ryanbourneuk) [reply ThunderRequest/RequestResponse.swift]: Potentially `preRedirectResponse`? Or `originalResponse`?

File: ThunderRequest/UIAlertController+ErrorRecovery.swift [hunk 42]
```diff
@@ -0,0 +1,109 @@
...
+        )
+        
+        switch error {
+        case var anyRecoverableError as CustomisableRecoverableError:
+            
+            if anyRecoverableError.options.isEmpty {
+                anyRecoverableError.add(option: ErrorRecoveryOption(title: "Dismiss", style: .cancel))
+            }
+            
+            anyRecoverableError.options.enumerated().forEach { (index, option) in
```
Comment (ryanbourneuk) [inline ThunderRequest/UIAlertController+ErrorRecovery.swift:86]: Following the same logic as the case block below, if the options are empty is it worth doing this? The above could be a guard instead & just return out before doing this :)

File: ThunderRequestTests/DownloadTests.swift [hunk 43]
```diff
@@ -0,0 +1,41 @@
...
+class DownloadTests: XCTestCase {
+    
+    let requestBaseURL = URL(string: "https://via.placeholder.com/")!
+    
+    func testDownloadSavesToDisk() {
+        
+        let requestController = RequestController(baseURL: requestBaseURL)
+        let finishExpectation = expectation(description: "App should correctly download body from server")
+        
+        requestController.download("500", progress: nil) { (response, url, error) in
```
Comment (ryanbourneuk) [inline ThunderRequestTests/DownloadTests.swift:22]: Is this making a network request? If so, this could fail if we're on the office WiFi / the service goes down.

File: ThunderRequestTests/UploadTests.swift [hunk 44]
```diff
@@ -0,0 +1,16 @@
...
+//  Created by Simon Mitchell on 14/12/2018.
+//  Copyright © 2018 threesidedcube. All rights reserved.
+//
+
+import UIKit
+import XCTest
+@testable import ThunderRequest
+
+class UploadTests: XCTestCase {
+
```
Comment (ryanbourneuk) [inline ThunderRequestTests/UploadTests.swift]: Empty tests

**Review by ryanbourneuk — COMMENTED** [review]
Looking good! A few comments throughout but nothing major. Lemme know what you think, beyond these I'm happy for this to be merged :)

File: .travis.yml (hunk 32 above)
Comment (simonmitchell) [reply .travis.yml:4]: Yeah 10.2 work (Not including travis yet) is up on a `swift_5` branch :) Will PR that once this one is done!

File: ThunderRequest/RequestController.swift (hunk 22 above)
Comment (simonmitchell) [reply ThunderRequest/RequestController.swift:111]: This is very true, I'm more just wary that if we remove this now (It has always been present in the past) we're going to get issues in countless apps that we may not pick up on immediately!

File: ThunderRequest/Data+MultipartFormElement.swift (hunk 35 above)
Comment (simonmitchell) [reply ThunderRequest/Data+MultipartFormElement.swift:13]: Is documented in the protocol declaration :D

File: ThunderRequest/RequestController+Auth.swift (hunk 37 above)
Comment (simonmitchell) [reply ThunderRequest/RequestController+Auth.swift:23]: This is called for every single request (And only internally), and so if no authenticator is provided we assume that requests don't need to be authenticated :) Would be a bit of a barrier to entry if users had to provide an `authenticator` object even if they were doing entirely un-authed requests!

File: ThunderRequest/CredentialStore.swift (hunk 33 above)
Comment (simonmitchell) [reply ThunderRequest/CredentialStore.swift:82]: Aye, doesn't need it to call `Update` as all the query seems to need is the service, account and class... all of which are present in the return from `keychainQueryWith`

File: ThunderRequest/UIAlertController+ErrorRecovery.swift (hunk 42 above)
Comment (simonmitchell) [reply ThunderRequest/UIAlertController+ErrorRecovery.swift:86]: I think we'd then be hiding errors from the user that the consumer of this API clearly wanted them to see. I wouldn't expect to call `UIAlertController(error: _someError)` and have it return a broken `UIAlertController` :)

File: ThunderRequestTests/DownloadTests.swift (hunk 43 above)
Comment (simonmitchell) [reply ThunderRequestTests/DownloadTests.swift:22]: It is, it's an integration test :) It 100% could fail, but not sure what else we can do about it? :/

**Review by ryanbourneuk — APPROVED** [review]
Superb work! Thank you for going through those comments too! Approving! :D
